### PR TITLE
Uniffi sync manager

### PR DESCRIPTION
**Pull request by bendk**
Based on #4634 

I made most of these changes mechanically.  I think we need someone who knows what these fields actually mean to double check that the changes make sense.

a-c PR: https://github.com/mozilla-mobile/android-components/pull/11256

### Pull Request checklist ###

- [ ] **Quality**: This PR builds and tests run cleanly
  - Note:
    - For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
    - If this pull request includes a breaking change, consider [cutting a new release](https://github.com/mozilla/application-services/blob/main/docs/howtos/cut-a-new-release.md) after merging.
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes a changelog entry in [CHANGES_UNRELEASED.md](../CHANGES_UNRELEASED.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [ ] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/main/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due dilligence applied in selecting them.

### CODE REVIEW

**Comment by lougeniaC64** [comment]
> I made most of these changes mechanically. I think we need someone who knows what these fields actually mean to double check that the changes make sense.
> 
@bendk Which fields are you referring to specifically? SyncParams??

**Comment by bendk** [comment]
> > I made most of these changes mechanically. I think we need someone who knows what these fields actually mean to double check that the changes make sense.
> 
> @bendk Which fields are you referring to specifically? SyncParams??

SyncParams and SyncResult,.  And both of them have several nested classes.

File: megazords/full/Cargo.toml [hunk 1]
```diff
@@ -11,11 +11,12 @@ crate-type = ["cdylib"]
 [dependencies]
 fxa-client = { path = "../../components/fxa-client" }
 logins = { path = "../../components/logins" }
+tabs = { path = "../../components/tabs/" }
```
Comment (mhammond) [inline megazords/full/Cargo.toml:14]: huh - why is this necessary and/or how did it work without it?

**Review by mhammond — APPROVED** [review]
This is awesome, great work Ben!

**Comment by mhammond** [comment]
> SyncParams and SyncResult,. And both of them have several nested types.

They look correct to me.  There are some minor semantic changes (eg, splitting success and error) which all make sense, but are obviously changes that need to carry through to a-c.

File: megazords/full/Cargo.toml (hunk 1 above)
Comment (bendk) [reply megazords/full/Cargo.toml:14]: I think because `sync_manager_ffi` depended on tabs so it was pulled in transitively.  But then `sync_manager` also depends on tabs, so I'm not exactly sure what's going on.  In any case, it seems good to explicitly list all the components that are in the megazord.

File: components/sync_manager/src/syncmanager.udl [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+};
+
+dictionary SyncResult {
+    ServiceStatus status;
+    record<DOMString, string> failures;
+    sequence<string> successful;
+    string persisted_state;
+    sequence<string>? declined;
+    timestamp? next_sync_allowed_at;
+    string? telemetry_json;
```
[the remark this answers is not in the document]
Comment (bendk) [reply components/sync_manager/src/syncmanager.udl:73]: I know, so close.  I think having external types would also work, since then this could just be a `SyncTelemetryPing` record.  I also thought about wrapping this class in a kt file that we define and giving it a `telemetry()` method, but it just didn't seem worth it yet.

**Comment by bendk** [comment]
Thanks for the review.  I removed the protobuf related stuff.  One last thing I want to do is add some documentation to the `SyncParams` and `SyncResults` types.  I started writing it, but then realized that there's several fields that I don't really get.  Can you tell me what I should replace the `TODO`s with in the [UDL file](https://github.com/mozilla/application-services/blob/220bbdbaa873f754f74a499c384680e76d06c218/components/sync_manager/src/syncmanager.udl)?

File: components/sync_manager/src/syncmanager.udl [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+    "PlacesError",
+    "AnyhowError",
+};
+
+dictionary SyncParams {
+    // Why are we performing this sync?
+    SyncReason reason;
+    // Which engines should we sync?
+    SyncEngineSelection engines;
+    // TODO
```
Comment (mhammond) [inline components/sync_manager/src/syncmanager.udl:25]: The changes that this sync should make to the "account global" list of enabled/disabled engines (for example, if the UI was used to change an engine's state since the last sync.

File: components/sync_manager/src/syncmanager.udl [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    record<DOMString, boolean> enabled_changes;
+    // Keys to encrypt/decrypt data from local database files.  These are
+    // separate from the key we use to encrypt the sync payload as a whole.
+    record<DOMString, string> local_encryption_keys;
+    // Authorization for the sync server
+    SyncAuthInfo auth_info;
+    // TODO
```
Comment (mhammond) [inline components/sync_manager/src/syncmanager.udl:32]: An opaque string, as returned in the previous sync's SyncResult and persisted to disk, or `null` if no such state is available. This includes information such as the list of engines previously enabled, certain server timestamps and GUIDs etc. If this value isn't correctly persisted and round-tripped, each sync may look like a "first sync".

File: components/sync_manager/src/syncmanager.udl [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+    string? persisted_state;
+    // TODO
+    DeviceSettings device_settings;
```
Comment (mhammond) [inline components/sync_manager/src/syncmanager.udl:35]: Information about the current device, such as its name, formfactor and FxA device ID.

File: components/sync_manager/src/syncmanager.udl [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+};
+
+dictionary SyncResult {
+    // Result from the sync server
+    ServiceStatus status;
+    // Engines that synced successfully
+    sequence<string> successful;
+    // Maps the names of engines that failed to sync to the reason why
+    record<DOMString, string> failures;
+    // TODO
```
Comment (mhammond) [inline components/sync_manager/src/syncmanager.udl:80]: State that should be persisted to disk and supplied to the `sync` method on the next sync.

File: components/sync_manager/src/syncmanager.udl [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+    string persisted_state;
+    // TODO
+    sequence<string>? declined;
```
Comment (mhammond) [inline components/sync_manager/src/syncmanager.udl:83]: The list of engines which are marked as "declined" (ie, disabled) on the sync server. The list of declined engines is global to the account rather than to the device. Apps should use this after every sync to update the local state (ie, to ensure that their Sync UI correctly reflects what engines are enabled and disabled), because these could change after every sync.
